### PR TITLE
Core components content fragment handling fixes

### PR DESCRIPTION
**Pull request by kaczymuczy**
## Description
Fixes for BB enabling the handling of Content Fragment core component
Required for: https://github.com/Cognifide/bobcat-aem-tests/pull/13
Part of https://github.com/Cognifide/bobcat/issues/373

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] My code follows the [code styleguide](https://github.com/Cognifide/bobcat/blob/master/CONTRIBUTING.md#styleguide) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultNumberInput.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+import com.cognifide.qa.bb.qualifier.PageObject;
+
+/**
+ * Default implementation of {@link NumberInput}
+ */
+@PageObject(css = Locators.NUMBERINPUT_CSS)
```
Comment (mkrzyzanowski) [inline bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultNumberInput.java]: If the locator is not used anywhere else, put it here directly.

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultPathBrowser.java [hunk 2]
```diff
@@ -52,8 +53,8 @@
   public void setValue(Object value) {
     input.clear();
     input.sendKeys(String.valueOf(value));
-
-    bobcatWait.until(elementToBeClickable(firstResult)).click();
+    bobcatWait.until(elementToBeClickable(firstResult));
+    input.sendKeys(Keys.ENTER);
```
Comment (mkrzyzanowski) [inline bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultPathBrowser.java]: Why the change from the previous implementation?

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultSelect.java [hunk 3]
```diff
@@ -26,22 +26,29 @@
...
+import com.cognifide.qa.bb.qualifier.CurrentScope;
 import com.cognifide.qa.bb.qualifier.PageObject;
+import com.cognifide.qa.bb.wait.BobcatWait;
+import com.google.inject.Inject;
 
 /**
  * Default implementation of {@link Select}
  */
-@PageObject(xpath = "//coral-select/..")
+@PageObject(css = Locators.SELECT_CSS)
```
Comment (mkrzyzanowski) [inline bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultSelect.java]: Same as above - if this is the only place where the selector is used, leave it here.

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultSelect.java [hunk 4]
```diff
@@ -26,22 +26,29 @@
...
+  @Inject
+  @CurrentScope
   private WebElement selectField;
 
-  @FindBy(css = Locators.LABEL_CSS)
+  @FindBy(xpath = Locators.ALTERNATE_LABEL_XPATH)
   private List<WebElement> label;
 
+  @Inject
+  private BobcatWait wait;
```
Comment (mkrzyzanowski) [inline bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultSelect.java]: Not used?

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultTextfield.java [hunk 5]
```diff
@@ -24,18 +24,21 @@
...
 
+import com.cognifide.qa.bb.qualifier.CurrentScope;
 import com.cognifide.qa.bb.qualifier.PageObject;
+import com.google.inject.Inject;
 
 /**
  * Default implementation of {@link Textfield}
  */
-@PageObject(css = Locators.FIELD_WRAPPER_CSS)
+@PageObject(css = Locators.TEXTFIELD_CSS)
```
Comment (mkrzyzanowski) [inline bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultTextfield.java]: Same comment regarding locator

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import com.cognifide.qa.bb.qualifier.PageObject;
+import com.cognifide.qa.bb.wait.BobcatWait;
+import com.cognifide.qa.bb.wait.TimingsBuilder;
+import com.google.inject.Inject;
+
+/**
+ * Default implementation of {@link PathBrowser}
+ */
+@PageObject(css = Locators.AUTOCOMPLETE_CSS)
+public class ContentFragmentPathBrowserImpl implements ContentFragmentPathBrowser {
```
Comment (mkrzyzanowski) [inline bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java]: The `DefaultXYZ` naming is more aligned with the rest fields

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    input.clear();
+    input.sendKeys(String.valueOf(value));
+    bobcatWait.until(elementToBeClickable(firstResult));
+    input.sendKeys(Keys.ENTER);
+
+    // Branching logic in tests is not OK, it's true. But this part handles an unpredictable part of
+    // the application. Sometimes a warning dialog appears and it needs to be handled.
+    if (bobcatWait
+        .tweak(new TimingsBuilder().explicitTimeout(1).build())
+        .ignoring(Stream.of(NoSuchElementException.class).collect(Collectors.toList()))
```
[the remark this answers is not in the document]
Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java]: I was testing this solution and thought that I might add more Exceptions later. But I didn't :D

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  @Override
+  public void setValue(Object value) {
+    input.clear();
+    input.sendKeys(String.valueOf(value));
+    bobcatWait.until(elementToBeClickable(firstResult));
+    input.sendKeys(Keys.ENTER);
+
+    // Branching logic in tests is not OK, it's true. But this part handles an unpredictable part of
+    // the application. Sometimes a warning dialog appears and it needs to be handled.
+    if (bobcatWait
```
[the remark this answers is not in the document]
Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java]: OK, will move to separate method.
The warning message doesn't always appear and I don't know yet what causes it to appear.
I know I'm ignoring the results of isConditionMet - it's a pretty sleep, nothing more, but it helps me avoid handling Selenium Exceptions

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultPathBrowser.java (hunk 2 above)
Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultPathBrowser.java]: The previous implementation didn't support external links, like to github.com - BB would click the first result suggestem by AEM, so a page within the AEM instance instead of github.com

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.cognifide.qa.bb.aem.core.component.dialog.dialogfields;
+
+import static org.openqa.selenium.support.ui.ExpectedConditions.*;
```
[the remark this answers is not in the document]
Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java]: Fixed

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java]: Fixed

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultNumberInput.java (hunk 1 above)
Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultNumberInput.java]: Fixed

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultSelect.java (hunk 3 above)
Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultSelect.java]: Fixed

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultSelect.java (hunk 4 above)
Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultSelect.java]: Fixed

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultTextfield.java (hunk 5 above)
Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultTextfield.java]: Fixed

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java (hunk 6 above)
Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/ContentFragmentPathBrowserImpl.java]: Fixed

File: bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultContentFragmentPathBrowser.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
+/*-
+ * #%L
+ * Bobcat
+ * %%
+ * Copyright (C) 2016 Cognifide Ltd.
```
Comment (mkrzyzanowski) [inline bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultContentFragmentPathBrowser.java]: Lies, it's 2019

Comment (kaczymuczy) [reply bb-aem-65/src/main/java/com/cognifide/qa/bb/aem/core/component/dialog/dialogfields/DefaultContentFragmentPathBrowser.java]: Fixed
